src/pd/client.rs

```diff
@@ -17,33 +17,93 @@ use std::time::Duration;
 use std::sync::Mutex;
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::thread;
+use std::collections::HashSet;
 use util::codec::rpc;
 use util::make_std_tcp_conn;
 
 use rand::{self, Rng};
 
-use kvproto::pdpb::{Request, Response};
+use kvproto::pdpb::{self, Request, Response};
 use kvproto::msgpb::{Message, MessageType};
 
-use super::{Result, PdClient};
+use super::{Result, PdClient, protocol};
 use super::metrics::*;
 
 const MAX_PD_SEND_RETRY_COUNT: usize = 100;
 const SOCKET_READ_TIMEOUT: u64 = 3;
 const SOCKET_WRITE_TIMEOUT: u64 = 3;
 
 const PD_RPC_PREFIX: &'static str = "/pd/rpc";
+const MSG_ID_VALIDATE: u64 = 0;
+
+// `validate_endpoints` validates pd members, make sure they are in the same cluster.
+// Notice that it ignores failed pd nodes.
+fn validate_endpoints(endpoints: &[String]) -> Result<()> {
+    if endpoints.is_empty() {
+        return Err(box_err!("empty PD list"));
+    }
+
+    let len = endpoints.len();
+    let mut endpoints_set = HashSet::with_capacity(len);
+
+    let mut sample = None;
+    for ep in endpoints {
+        if !endpoints_set.insert(ep) {
+            return Err(box_err!("a duplicate PD url {}", ep));
+        }
+
+        let mut stream = match rpc_connect(ep.as_str()) {
+            Ok(stream) => stream,
+            // Ignore failed pd node.
+            Err(_) => continue,
+        };
+
+        let mut req = protocol::new_request(MSG_ID_VALIDATE, pdpb::CommandType::GetPDMembers);
+        req.set_get_pd_members(pdpb::GetPDMembersRequest::new());
+        let (id, mut resp) = match send_msg(&mut stream, MSG_ID_VALIDATE, &req) {
+            Ok((id, resp)) => (id, resp),
+            // Ignore failed pd node.
+            Err(_) => continue,
+        };
+
+        if id != MSG_ID_VALIDATE {
+            return Err(box_err!("PD response msg_id not match, want {}, got {}",
+                                MSG_ID_VALIDATE,
+                                id));
+        }
+
+        let mut members = resp.take_get_pd_members().take_members().into_vec();
+        if members.len() < len {
+            return Err(box_err!("inconsistent PD list, there is an invalid PD url"));
+        }
+
+        // Check all fields.
+        members.sort_by(|a, b| a.get_name().cmp(b.get_name()));
+        if let Some(ref sample) = sample {
+            if *sample != members {
+                return Err(box_err!("inconsistent PD list, expect: {:?}, got: {:?}",
+                                    sample,
+                                    members));
+            }
+        } else {
+            sample = Some(members);
+        }
+    }
+
+    Ok(())
```

Comment:
Checking cluster ID should be enough, but additional checks prevent ID collision.